packages/fee-abstraction/src/storage.rs

```diff
@@ -0,0 +1,388 @@
+/// The caller must ensure proper authorization before calling this function.
+pub fn set_allowed_fee_token(e: &Env, token: &Address, allowed: bool) {
+    let count_key = FeeAbstractionStorageKey::Count;
+    let mut count: u32 = e.storage().instance().get(&count_key).unwrap_or(0);
+
+    let token_index_key = FeeAbstractionStorageKey::TokenIndex(token.clone());
+    let existing_index: Option<u32> = e.storage().persistent().get(&token_index_key);
+
+    if allowed {
+        if existing_index.is_some() {
+            // Trying to allow an already-allowed token.
+            panic_with_error!(e, FeeAbstractionError::FeeTokenAlreadyAllowed);
+        }
+
+        // Assign new index at the end.
+        e.storage().persistent().set(&FeeAbstractionStorageKey::Token(count), token);
+        e.storage().persistent().set(&token_index_key, &count);
+
+        // Increment count.
+        count = count
+            .checked_add(1)
+            .unwrap_or_else(|| panic_with_error!(e, FeeAbstractionError::TokenCountOverflow));
+        e.storage().instance().set(&count_key, &count);
+    } else {
+        let remove_index = existing_index
+            .unwrap_or_else(|| panic_with_error!(e, FeeAbstractionError::FeeTokenNotAllowed));
+
+        // Can't underflow, it would've been caught be the above panic_with_error
+        let last_index = count - 1;
+        let last_key = FeeAbstractionStorageKey::Token(last_index);
+
+        // Swap and pop
+        if remove_index != last_index {
+            // Move last token into the removed slot.
+            let last_token: Address =
+                e.storage().persistent().get(&last_key).expect("last token to be present");
+
+            e.storage()
+                .persistent()
+                .set(&FeeAbstractionStorageKey::Token(remove_index), &last_token);
+
+            // Update moved token's index mapping.
+            e.storage()
+                .persistent()
+                .set(&FeeAbstractionStorageKey::TokenIndex(last_token.clone()), &remove_index);
+        }
+
+        // Remove last index entry.
+        e.storage().persistent().remove(&last_key);
+
+        // Remove mapping for the removed token.
+        e.storage().persistent().remove(&token_index_key);
+
+        count -= 1;
+        e.storage().instance().set(&count_key, &count);
+    }
+
+    emit_fee_token_allowlist_updated(e, token, allowed);
+}
+
+/// Check if a token is allowed for fee payment.
+///
+/// If the allowlist is disabled (no fee tokens added), all tokens are
+/// considered allowed. If the allowlist is enabled (at least one fee token is
+/// added), only explicitly allowed tokens are permitted.
+///
+/// # Arguments
+///
+/// * `e` - Access to Soroban environment.
+/// * `token` - The token contract address to check.
+///
+/// # Errors
+///
+/// * [`FeeAbstractionError::FeeTokenNotAllowed`] - If the token is not allowed.
+pub fn check_allowed_fee_token(e: &Env, token: &Address) {
```

Comment:
hmm, what is the reason behind not returning a `bool` but directly panicking? From what I've seen, there is no function provided to the users to first check if the token is allowed or not without the panicking behavior.

Reply:
The allow list is enabled only if there's at least one token that was allowed. And this is a convenience function that checks both parts: the allow list is enabled and the provided token is allowed. If it returns a `bool`, it won't be clear which condition is not met.

Reply:
@ozgunozerk that's right, check out now